src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/service/BookingTimelineService.kt

```diff
@@ -64,19 +66,63 @@ class BookingTimelineService(
           val daysRemand = booking.getOrZero(AdjustmentType.REMAND, it.sentencedAt) - daysRemandNoLongerApplicable
-          if (workingRange.end.plusDays(daysRemand.toLong()).isBefore(it.sentencedAt.minusDays(1))) {
-            throw BookingTimelineGapException("There is a gap between sentences $previousSentence AND $it")
+          if (sentenceRange.end.plusDays(daysRemand.toLong()).isBefore(it.sentencedAt.minusDays(1))) {
+            throw BookingTimelineGapException(
+              "There is a gap between sentence sentence at ${previousSentence.sentencedAt}" +
+                " release of ${previousSentence.sentenceCalculation.adjustedReleaseDate}" +
+                " AND sentence at ${it.sentencedAt} release of ${it.sentenceCalculation.adjustedReleaseDate}"
+            )
           } else {
-            workingRange = LocalDateRange.of(workingRange.start, itRange.end)
+            sentenceRange = LocalDateRange.of(sentenceRange.start, itRange.end)
           }
         }
       }
       previousSentence = it
       readjustDates(it, daysAwardedServed, daysRemandNoLongerApplicable)
     }
+
+    validateRemandPeriodsOverlapping(booking)
     return booking
   }
 
+  private fun validateRemandPeriodsOverlapping(booking: Booking) {
+    val remandPeriods = booking.get(AdjustmentType.REMAND)
+    if (remandPeriods.isNotEmpty()) {
+      val remandRanges = remandPeriods.map { LocalDateRange.of(it.fromDate, it.toDate) }
+      val sentenceRanges = booking.getAllExtractableSentences().map { LocalDateRange.of(it.sentencedAt, it.sentenceCalculation.adjustedReleaseDate) }
+
+      val allRanges = (remandRanges + sentenceRanges).sortedBy { it.start }
+      var totalRange: LocalDateRange? = null
+      var previousRangeIsRemand: Boolean? = null
+      var previousRange: LocalDateRange? = null
+
+      allRanges.forEach {
+        val isRemand = remandRanges.any() { sentenceRange -> sentenceRange === it }
+        if (totalRange == null && previousRangeIsRemand == null) {
+          totalRange = it
+          previousRangeIsRemand = isRemand
+          previousRange = it
+        } else {
+          if (it.isConnected(totalRange) &&
+            (previousRangeIsRemand!! || isRemand)
+          ) {
+            // Remand overlaps
+            if (previousRangeIsRemand!! && isRemand) {
+              throw RemandPeriodOverlapsWithRemandException("Remand of range ${previousRange!!} overlaps with remand of range $it")
+            } else {
+              throw RemandPeriodOverlapsWithSentenceException("${if (previousRangeIsRemand!!) "Remand" else "Sentence"} of range ${previousRange!!} overlaps with ${if (isRemand) "remand" else "sentence"} of range $it")
+            }
+          }
+          if (it.end.isAfter(totalRange!!.end)) {
+            totalRange = LocalDateRange.of(totalRange!!.start, it.end)
+          }
+          previousRangeIsRemand = isRemand
+          previousRange = it
+        }
+      }
+    }
+  }
+
```

Comment:
could this whole section be changed to something like this? I think logica,lly it's the same and a bit easier on the eye
```

      allRanges.forEach {
        val isRemand = remandRanges.any { sentenceRange -> sentenceRange === it }
        if (totalRange == null && previousRangeIsRemand == null) {
          totalRange = it
        } else if (it.isConnected(totalRange) && (previousRangeIsRemand!! || isRemand)) {
          // Remand overlaps
          if (previousRangeIsRemand!! && isRemand) {
            throw RemandPeriodOverlapsWithRemandException("Remand of range ${previousRange!!} overlaps with remand of range $it")
          } else {
            throw RemandPeriodOverlapsWithSentenceException("${if (previousRangeIsRemand!!) "Remand" else "Sentence"} of range ${previousRange!!} overlaps with ${if (isRemand) "remand" else "sentence"} of range $it")
          }
        } else if (it.end.isAfter(totalRange!!.end)) {
          totalRange = LocalDateRange.of(totalRange!!.start, it.end)
        }
        previousRangeIsRemand = isRemand
        previousRange = it
      }
```

Reply:
Yea done, good spot!